### DISCUSSION TITLE
Testing Node-RED Dashboard on Raspberry pi pico W with pico display 2.0

### DISCUSSION THREAD

**Discussion by kitazaki** [phoddie/node-red-mcu#119] [Q&A] [answered]
I tested Raspberry pi pico W with pico display 2.0.
It works fine with Moddable SDK both examples piu/balls and piu/images.

![IMG_1309](https://github.com/phoddie/node-red-mcu/assets/23477407/38b8a0d2-4c68-4115-afc4-a10f62a933ed)
![IMG_1310](https://github.com/phoddie/node-red-mcu/assets/23477407/f5595fb3-a629-4079-9a2b-eaa13d212f8a)

I tried Node-RED Dashboard on it, but it doesn't work and stopped working after turned it on soon.
Any ideas for working?

[command]
% mcconfig -d -m -p pico/pico_display_2

[Node-RED flow]
<img width="439" alt="スクリーンショット 2023-06-24 11 07 26" src="https://github.com/phoddie/node-red-mcu/assets/23477407/12587ff6-4ac7-4db6-ad59-dd70b14dbbef">

[xs-bug]
<img width="1110" alt="スクリーンショット 2023-06-24 11 06 14" src="https://github.com/phoddie/node-red-mcu/assets/23477407/999c82c7-0350-4350-815d-7fcd0305b53b">

[flows.json]
```json
[
    {
        "id": "70caf9a6d711f6a7",
        "type": "tab",
        "label": "フロー 1",
        "disabled": false,
        "info": "",
        "env": [],
        "_mcu": {
            "mcu": true
        }
    },
    {
        "id": "de0fcfd6a81e1634",
        "type": "ui_tab",
        "name": "home",
        "icon": "dashboard",
        "disabled": false,
        "hidden": false,
        "_mcu": {
            "mcu": false
        }
    },
    {
        "id": "21ef63c93608c616",
        "type": "ui_base",
        "theme": {
            "name": "theme-light",
            "lightTheme": {
                "default": "#0094CE",
                "baseColor": "#0094CE",
                "baseFont": "-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Oxygen-Sans,Ubuntu,Cantarell,Helvetica Neue,sans-serif",
                "edited": true,
                "reset": false
            },
            "darkTheme": {
                "default": "#097479",
                "baseColor": "#097479",
                "baseFont": "-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Oxygen-Sans,Ubuntu,Cantarell,Helvetica Neue,sans-serif",
                "edited": false
            },
            "customTheme": {
                "name": "Untitled Theme 1",
                "default": "#4B7930",
                "baseColor": "#4B7930",
                "baseFont": "-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Oxygen-Sans,Ubuntu,Cantarell,Helvetica Neue,sans-serif"
            },
            "themeState": {
                "base-color": {
                    "default": "#0094CE",
                    "value": "#0094CE",
                    "edited": false
                },
                "page-titlebar-backgroundColor": {
                    "value": "#0094CE",
                    "edited": false
                },
                "page-backgroundColor": {
                    "value": "#fafafa",
                    "edited": false
                },
                "page-sidebar-backgroundColor": {
                    "value": "#ffffff",
                    "edited": false
                },
                "group-textColor": {
                    "value": "#1bbfff",
                    "edited": false
                },
                "group-borderColor": {
                    "value": "#ffffff",
                    "edited": false
                },
                "group-backgroundColor": {
                    "value": "#ffffff",
                    "edited": false
                },
                "widget-textColor": {
                    "value": "#111111",
                    "edited": false
                },
                "widget-backgroundColor": {
                    "value": "#0094ce",
                    "edited": false
                },
                "widget-borderColor": {
                    "value": "#ffffff",
                    "edited": false
                },
                "base-font": {
                    "value": "-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Oxygen-Sans,Ubuntu,Cantarell,Helvetica Neue,sans-serif"
                }
            },
            "angularTheme": {
                "primary": "indigo",
                "accents": "blue",
                "warn": "red",
                "background": "grey",
                "palette": "light"
            }
        },
        "site": {
            "name": "Node-RED ダッシュボード",
            "hideToolbar": "false",
            "allowSwipe": "false",
            "lockMenu": "false",
            "allowTempTheme": "true",
            "dateFormat": "YYYY/MM/DD",
            "sizes": {
                "sx": 48,
                "sy": 48,
                "gx": 6,
                "gy": 6,
                "cx": 6,
                "cy": 6,
                "px": 0,
                "py": 0
            }
        }
    },
    {
        "id": "28c093aa4ff3c8bb",
        "type": "ui_group",
        "name": "default",
        "tab": "de0fcfd6a81e1634",
        "order": 1,
        "disp": true,
        "width": "6",
        "collapse": false,
        "className": "",
        "_mcu": {
            "mcu": false
        }
    },
    {
        "id": "dae37bbf6c9c4907",
        "type": "ui_text",
        "z": "70caf9a6d711f6a7",
        "group": "28c093aa4ff3c8bb",
        "order": 2,
        "width": 0,
        "height": 0,
        "name": "",
        "label": "text",
        "format": "{{msg.payload}}",
        "layout": "row-spread",
        "className": "",
        "style": false,
        "font": "",
        "fontSize": 16,
        "color": "#000000",
        "_mcu": {
            "mcu": true
        },
        "x": 330,
        "y": 160,
        "wires": []
    },
    {
        "id": "7c52637745382047",
        "type": "inject",
        "z": "70caf9a6d711f6a7",
        "name": "",
        "props": [
            {
                "p": "payload"
            },
            {
                "p": "topic",
                "vt": "str"
            }
        ],
        "repeat": "1",
        "crontab": "",
        "once": false,
        "onceDelay": 0.1,
        "topic": "",
        "payload": "",
        "payloadType": "date",
        "_mcu": {
            "mcu": true
        },
        "x": 150,
        "y": 100,
        "wires": [
            [
                "95e009c874f15a3f",
                "dae37bbf6c9c4907"
            ]
        ]
    },
    {
        "id": "95e009c874f15a3f",
        "type": "debug",
        "z": "70caf9a6d711f6a7",
        "name": "debug 2",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "false",
        "statusVal": "",
        "statusType": "auto",
        "_mcu": {
            "mcu": true
        },
        "x": 340,
        "y": 100,
        "wires": []
    }
]
```

**Comment by phoddie** [comment]
@kitazaki – it is great that you are trying this. Thank you.

One question for you: did you try the same flow on an ESP32 with a display? That would be helpful to know, since it would confirm the flow works correctly with the MCU Edition runtime. If you don't have an ESP32 with a display, you could also try the simulator.

As you can understand, the Node-RED dashboard requires more memory. It was a bit of a challenge to make it work on ESP32, which has more RAM than Pico. So, we probably have some work to do here to get it working. Please have some patience as we try.

I will take a look in the next day or two and share some suggestions and/or results.

**Comment by kitazaki** [comment]
@phoddie 

This flow works fine on M5Stack(esp32) and simulator(macOS).

![m5stack](https://github.com/phoddie/node-red-mcu/assets/23477407/872310e9-81d0-420f-b1bd-46a677829fe6)
![スクリーンショット 2023-06-24 14 07 00](https://github.com/phoddie/node-red-mcu/assets/23477407/f5555e38-ce6b-4a6f-9964-fff9e93aead3)

**Reply by phoddie** [reply]
Thanks for confirming that.

I did some experimentation. The good news is that I am pretty confident that this can work. There is plenty of memory available.

On ESP32, I changed the creation in `manifest_host.json` to be much smaller and it still runs:

```json
		"esp32": {
			"creation": {
				"static": 65536,
				"chunk": {
					"initial": 1024,
					"incremental": 1024
				},
				"heap": {
					"initial": 64,
					"incremental": 64
				},
				"keys": {
					"available": 8
				}
			}
		}
```

I also reduced the configuration options of the Piu application instance in main.js:

```js
new REDApplication(model, { commandListLength:1024, displayListLength:1024, touchCount:1, pixels: 240 * 2 });
```

These settings are unrealistic for most projects, but I am using them to help understand where this is failing.

On the Pico, with these settings it still fails. It fails in the call to the `REDApplication` constructor. Instantiating Piu objects is one of the deepest uses of the stack we have. I think it is failing because the native stack is overflowing. I can't prove that easily. The native stack on Raspberry Pi Pico is configured to 4 KB (this is usually enough). On ESP32, the main stack is at least 8 KB.

Unfortunately, I was unsuccessful at changing the Pico stack. At first it seemed it would be enough to set `PICO_STACK_SIZE` but that seems untrue. The right way seems to modify the linker script. I tried that quickly but it didn't work. Maybe you will have more luck? The Moddable SDK has its own linker script for the Pico, and the [relevant bits seem to be here](https://github.com/Moddable-OpenSource/moddable/blob/c58c6e9910ae270fd8f7c7a9758de367331e94f8/build/devices/pico/config/xsproj.ld#L257-L263).

That's all I have time for today to look into this. Perhaps you will have some insight based on these notes.

**Reply by phoddie** [reply]
Actually.... the native stack at 4 KB is big enough. The only change required is to adjust the creation parameters for Pico in `manifest_host.json` to those shown above for ESP32. 

![IMG_3499](https://github.com/phoddie/node-red-mcu/assets/6856458/cd0a963e-6d5b-4bff-b3f4-95397fdd5a64)

We'll need to do some more work to find optimal settings here (and understand why it is silently crashing with the current defaults).  I would also like to know how to adjust the native stack because eventually it will be necessary. But, it can work. ;)

**Comment by kitazaki** [comment, marked as answer]
This works great! Thanks!
![IMG_1316](https://github.com/phoddie/node-red-mcu/assets/23477407/e25440c6-784e-41e6-9b2c-c10b65549c1b)

I only change the Creation Params.

[Creation Params]
```json
"pico": {
			"creation": {
				"static": 65536,
				"chunk": {
					"initial": 1024,
					"incremental": 1024
				},
				"heap": {
					"initial": 64,
					"incremental": 64
				},
				"keys": {
					"available": 8
				}
			}
		}
```

It's same Configuration as ESP32

[Configuration]
<img width="426" alt="スクリーンショット 2023-06-25 8 24 19" src="https://github.com/phoddie/node-red-mcu/assets/23477407/1d5500d1-bd8e-477e-8983-f57ee050804d">

**Reply by phoddie** [reply]
Glad this is working for you now! I think probably we need some different defaults on the Pico, since it has less memory than ESP32. Of course, some projects may need to adjust the defaults but it would be nice for basic dashboard projects to work "out of the box."
